File: src/algorithms/AlgorithmMinCircle.cpp
#include <algorithm>
#include <limits>

#include "../include/algorithms/AlgorithmMinCircle.h"

AlgorithmMinCircle::~AlgorithmMinCircle()
{}

Circle2D AlgorithmMinCircle::naiveAlgorithm(std::vector<Point2D> const &points) const
{
    // General case
    //std::cout << "General case\n";
    Circle2D minimumCircle;
    for (auto p = begin(points); p != end(points); ++p)
    {
        for (auto q = begin(points); q != end(points); ++q)
        {
            if (p != q)
            {
                minimumCircle.center.x = (p->x + q->x) * 0.5,
                minimumCircle.center.y = (p->y + q->y) * 0.5,
                minimumCircle.radius = p->distanceTo(*q) * 0.5;

                if (minimumCircle.containsAllPoints(begin(points), end(points)))
                {
                    //std::cout << "OK\n";
                    return minimumCircle;
                }
            }
        }
    }

    //std::cout << "Triangle case\n";

    // Triangle case
    minimumCircle.radius = std::numeric_limits<double>::max();
    for (auto p = begin(points); p != end(points); ++p)
    {
        for (auto q = begin(points); q != end(points); ++q)
        {
            for (auto r = begin(points); r != end(points); ++r)
            {
                if (p != q && q != r && p != r && !Point2D::areColinear(*p, *q, *r))
                {
                    Circle2D const CIRCUMSCRIBED_CIRCLE{*p, *q, *r};
                    if (CIRCUMSCRIBED_CIRCLE.radius < minimumCircle.radius && CIRCUMSCRIBED_CIRCLE.containsAllPoints(begin(points), end(points))) {
                        minimumCircle.center.x = CIRCUMSCRIBED_CIRCLE.center.x;
                        minimumCircle.center.y = CIRCUMSCRIBED_CIRCLE.center.y;
                        minimumCircle.radius = CIRCUMSCRIBED_CIRCLE.radius;
                    }
                }
            }
        }
    }

    return minimumCircle;
}

Circle2D AlgorithmMinCircle::minidisk(std::vector<Point2D> &points) const
{
    std::vector<Point2D> boundary;
    boundary.reserve(3);

    return bMinidisk(points, boundary);
}

Circle2D AlgorithmMinCircle::bMinidisk(std::vector<Point2D> &points, std::vector<Point2D> &boundary) const
{
    Circle2D minimumCircle;
    if (boundary.size() == 3)
    {
        minimumCircle = Circle2D{boundary[0], boundary[1], boundary[2]};
    }
    else if (points.empty() && boundary.size() == 2)
    {
        minimumCircle = Circle2D{boundary[0], boundary[1]};
    }
    else if (points.size() == 1 && boundary.empty())
    {
        minimumCircle = Circle2D{points[0].x, points[0].y, 0.0};
    }
    else if (points.size() == 1 && boundary.size() == 1)
    {
        minimumCircle = Circle2D{points[0], boundary[0]};
    }
    else
    {
        Point2D p = *points.erase(points.end() - 1);
        minimumCircle = bMinidisk(points, boundary);

        if (minimumCircle.containsPoint(p))
        {
            boundary.push_back(p);
            minimumCircle = bMinidisk(points, boundary);

            boundary.erase(points.end() - 1);
            points.push_back(p);
        }
    }

    return minimumCircle;
}

File: java/src/com/upmc/alogirthms/AlgorithmMinCircle.java
package com.upmc.alogirthms;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import com.upmc.geom.Circle2D;
import com.upmc.geom.Point2D;

public class AlgorithmMinCircle {
	private Random rand = new Random();
	
    public Circle2D naiveAlgorithm(final List<Point2D> points) {
    	// One point = nul circle
    	if (points.size() == 1) {
    		return new Circle2D(points.get(0).x, points.get(0).y, 0.0);
    	}
    	
    	Circle2D minimumCircle = null;
    	
    	// General case
		for (final Point2D p : points) {
			for (final Point2D q : points) {
				if (p != q) {
					minimumCircle = new Circle2D(p, q);
					if (minimumCircle.containsAllPoints(points)) {
						return minimumCircle;
					}
				}
			}
		}

		// Triangle case
		minimumCircle = new Circle2D(0.0, 0.0, Double.MAX_VALUE);
		for (final Point2D p : points) {
			for (final Point2D q : points) {
				for (final Point2D r : points) {
					if (p != q && q != r && p != r && !Point2D.areColinear(p, q, r)) {
						Circle2D circumscribedCircle = new Circle2D(p, q, r);
						if (circumscribedCircle.radius < minimumCircle.radius && circumscribedCircle.containsAllPoints(points)) {
							minimumCircle = circumscribedCircle;
						}
					}
				}
			}
		}

		return minimumCircle;
    }
    
    public Circle2D minidisk(final List<Point2D> points) {
		return bMinidisk(points, new ArrayList<Point2D>());
    }
	
    private Circle2D bMinidisk(final List<Point2D> points, final List<Point2D> boundary) {
    	Circle2D minimumCircle = null;
		
		if (boundary.size() == 3) {
			minimumCircle = new Circle2D(boundary.get(0), boundary.get(1), boundary.get(2));
		}
		else if (points.isEmpty() && boundary.size() == 2) {
			minimumCircle = new Circle2D(boundary.get(0), boundary.get(1));
		}
		else if (points.size() == 1 && boundary.isEmpty()) {
			minimumCircle = new Circle2D(points.get(0).x, points.get(0).y, 0.0);
		}
		else if (points.size() == 1 && boundary.size() == 1) {
			minimumCircle = new Circle2D(points.get(0), boundary.get(0));
		}
		else {
			Point2D p = points.remove(rand.nextInt(points.size()));
			minimumCircle = bMinidisk(points, boundary);
			
			if (minimumCircle != null && !minimumCircle.containsPoint(p)) {
				boundary.add(p);
				minimumCircle = bMinidisk(points, boundary);
				boundary.remove(p);
				points.add(p);
			}
		}
				
		return minimumCircle;
    }
}
